lib/srv/sess.go

```diff
@@ -52,77 +56,160 @@ func (r *sessionRegistry) Close() {
+			events.SessionEventID:  string(ctx.session.id),
+			events.EventLogin:      ctx.login,
+			events.EventUser:       ctx.teleportUser,
+			events.LocalAddr:       ctx.conn.LocalAddr().String(),
+			events.RemoteAddr:      ctx.conn.RemoteAddr().String(),
+			events.SessionServerID: ctx.srv.ID(),
+		})
+		ctx.Infof("joining session: %v", ctx.session.id)
+		_, err := ctx.session.join(ch, req, ctx)
 		return trace.Wrap(err)
 	}
-
+	// session not found? need to create one. start by getting/generating an ID for it
+	sid, found := ctx.getEnv(sshutils.SessionEnvVar)
+	if !found {
+		sid = string(rsession.NewID())
+		ctx.setEnv(sshutils.SessionEnvVar, sid)
+	}
 	// This logic allows concurrent request to create a new session
 	// to fail, what is ok because we should never have this condition
-	sess, err := newSession(sid, s, ctx)
+	sess, err := newSession(rsession.ID(sid), s, ctx)
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	if err := sess.startShell(sconn, ch, ctx); err != nil {
-		defer sess.Close()
-		return trace.Wrap(err)
-	}
+	ctx.session = sess
 	s.addSession(sess)
-	ctx.Infof("created session: %v", sess.id)
+	ctx.Infof("ssh.joinShell created new session %v", sid)
 
+	if err := sess.startShell(ch, ctx); err != nil {
+		sess.Close()
+		return trace.Wrap(err)
+	}
 	return nil
 }
 
-func (s *sessionRegistry) leaveShell(sid, pid rsession.ID) error {
+// leaveShell remvoes a given party from this session
+func (s *sessionRegistry) leaveShell(party *party) error {
+	sess := party.s
 	s.Lock()
 	defer s.Unlock()
 
-	sess, found := s.findSession(sid)
-	if !found {
-		return trace.NotFound("session %v not found", sid)
-	}
-	if err := sess.leave(pid); err != nil {
+	// remove from in-memory representation of the session:
+	if err := sess.removeParty(party); err != nil {
 		return trace.Wrap(err)
 	}
-	if len(sess.parties) != 0 {
-		return nil
-	}
-	log.Infof("last party left %v, removing from server", sess)
-	delete(s.sessions, sess.id)
-	if err := sess.Close(); err != nil {
-		log.Errorf("failed to close: %v", err)
-		return err
+
+	// emit an audit event
+	s.srv.EmitAuditEvent(events.SessionLeaveEvent, events.EventFields{
+		events.SessionEventID:  string(sess.id),
+		events.EventUser:       party.user,
+		events.SessionServerID: party.serverID,
+	})
+
+	// this goroutine runs for a short amount of time only after a session
+	// becomes empty (no parties). It allows session to "linger" for a bit
+	// allowing parties to reconnect if they lost connection momentarily
+	lingerAndDie := func() {
+		lingerTTL := sess.GetLingerTTL()
+		if lingerTTL > 0 {
+			time.Sleep(lingerTTL)
+		}
+		// not lingering anymore? someone reconnected? cool then... no need
+		// to die...
+		if !sess.isLingering() {
+			log.Infof("[session.registry] session %v becomes active again", sess.id)
+			return
+		}
+		log.Infof("[session.registry] session %v to be garbage collected", sess.id)
+
+		// no more people left? Need to end the session!
+		s.Lock()
```

Comment:
just a suggestion: move this delete to a separate function to protect ourselves from deadlocks on panic